src/Adapter/Order/Refund/OrderProductRemover.php

```diff
@@ -231,4 +232,35 @@ private function deleteSpecificPrice(
             $specificPrice->delete();
         }
     }
+
+    /**
+     * This method removes cart rules which was applied to specific product when that product is being deleted.
+     *
+     * @param Order $order
+     * @param OrderDetail $orderDetail
+     *
+     * @return void
+     */
+    private function removeSpecificProductCartRules(
```

Comment:
note- this change was needed because of specific_product (the cartrule->reduction_product value) and product_restriction features mixing. It seems that only in BO it didn't check of $cartRule->reduction_product (and only checked restricted products) and while in old tests these 2 were mixed it used to work by chance.

The following test failed without this: `./vendor/bin/behat -c tests/Integration/Behaviour/behat.yml -s order --tags order-cart-rules` - `Add product linked to a cart rule to an existing Order without invoice with free shipping and new invoice And remove this product`.
The scenario name does not state it, but the main idea was that when removed specific product, the cart rule related to that product was not being removed. (because as I mentioned only restricted products were checked, while reduction_product should also be checked). In Front office it seems reduction_product is checked correctly.